src/cmakeTaskProvider.ts

```diff
@@ -118,64 +130,103 @@ class CustomBuildTaskTerminal implements vscode.Pseudoterminal, proc.OutputConsu
     public get onDidClose(): vscode.Event<number> {
         return this.closeEmitter.event;
     }
-    private endOfLine: string = "\r\n";
 
     constructor(private command: string, private defaultTargets: string[], private definedTargets?: string[], private options: { cwd?: string ; environment?: Environment } = {}, private cmakeDriver?: CMakeDriver) {
     }
 
     output(line: string): void {
-        this.writeEmitter.fire(line + this.endOfLine);
+        this.writeEmitter.fire(line + endOfLine);
     }
 
     error(error: string): void {
-        this.writeEmitter.fire(error + this.endOfLine);
+        this.writeEmitter.fire(error + endOfLine);
     }
 
     async open(_initialDimensions: vscode.TerminalDimensions | undefined): Promise<void> {
+        if (this.command !== CommandType.build && this.command !== CommandType.config && this.command !== CommandType.install && this.command !== CommandType.test) {
+            this.writeEmitter.fire(localize("command.not.recognized", '{0} is not a recognized command.', `"${this.command}"`) + endOfLine);
+            this.closeEmitter.fire(-1);
+            return;
+        }
         // At this point we can start using the terminal.
-        this.writeEmitter.fire(localize("starting.build", "Starting build...") + this.endOfLine);
-        await this.doBuild();
+        switch (this.command) {
+            case CommandType.build:
+                await this.runBuildTask();
+                break;
+            case CommandType.config:
+                await this.runConfigTask();
+                break;
+            case CommandType.install:
+                await this.runInstallTask();
+                break;
+            case CommandType.test:
+                await this.runTestTask();
+                break;
+        }
+
     }
 
     close(): void {
         // The terminal has been closed. Shutdown the build.
     }
 
-    private async doBuild(): Promise<any> {
-        if (this.command !== CommandType.build) {
-            this.writeEmitter.fire(localize("not.a.build.command", '{0} is not a recognized build command.', `"${this.command}"`) + this.endOfLine);
-            this.closeEmitter.fire(-1);
-            return;
-        }
-        let buildCommand: proc.BuildCommand | null;
+    private async runBuildTask(): Promise<any> {
+        let command: proc.BuildCommand | null;
         let cmakePath: string = "CMake.EXE";
         let args: string[] = [];
 
         if (this.cmakeDriver) {
-            buildCommand = await this.cmakeDriver.getCMakeBuildCommand(this.definedTargets ? this.definedTargets : this.defaultTargets);
-            if (buildCommand) {
-                cmakePath = buildCommand.command;
-                args = buildCommand.args ? buildCommand.args : [];
-                this.options.environment = buildCommand.build_env;
+            if (await this.cmakeDriver.checkNeedsReconfigure() || !isSubsetTarget(this.defaultTargets, this.definedTargets)) {
```

Comment:
I don't think we should force a configure if the defined target is not a subset.  I think just checking for reconfigure is good enough.

Reply:
(and then you can delete `isSubsetTarget`)